File: frontend/types/job.ts
import type { Speaker } from './speaker'

export type JobStatus =
  | 'UPLOADED'
  | 'PREPROCESSING'
  | 'VAD'
  | 'DIARIZING'
  | 'EMBEDDING'
  | 'CLUSTERING'
  | 'SEPARATING'
  | 'TRANSCRIBING'
  | 'ANALYZING'
  | 'RECONSTRUCTING'
  | 'COMPLETED'
  | 'FAILED'

export const JOB_STATUS_LABELS: Record<JobStatus, string> = {
  UPLOADED: 'Uploaded',
  PREPROCESSING: 'Preprocessing',
  VAD: 'Voice Detection',
  DIARIZING: 'Speaker Diarization',
  EMBEDDING: 'Voice Embeddings',
  CLUSTERING: 'Clustering',
  SEPARATING: 'Separating',
  TRANSCRIBING: 'Transcribing',
  ANALYZING: 'Analyzing',
  RECONSTRUCTING: 'Reconstructing',
  COMPLETED: 'Completed',
  FAILED: 'Failed',
}

export const PIPELINE_STAGES: Array<{ key: string; label: string; description: string }> = [
  { key: 'PREPROCESSING', label: 'Noise Reduction',     description: 'Cleaning audio and removing background noise' },
  { key: 'VAD',           label: 'Voice Detection',     description: 'Detecting regions of human speech' },
  { key: 'DIARIZING',     label: 'Speaker Diarization', description: 'Identifying who speaks when' },
  { key: 'EMBEDDING',     label: 'Voice Embeddings',    description: 'Creating unique voice fingerprints' },
  { key: 'CLUSTERING',    label: 'Clustering',          description: 'Grouping speech segments by speaker' },
  { key: 'SEPARATING',    label: 'Speech Separation',   description: 'Isolating each speaker\'s audio' },
  { key: 'TRANSCRIBING',  label: 'Transcription',       description: 'Converting speech to text' },
  { key: 'ANALYZING',     label: 'Analysis',            description: 'Computing speaker statistics' },
  { key: 'RECONSTRUCTING',label: 'Reconstruction',      description: 'Generating final audio files' },
]

export interface Job {
  id: string
  user_id: string
  original_filename: string
  original_s3_key: string
  processed_s3_key?: string
  status: JobStatus
  pipeline_stage?: string
  progress: number
  duration_seconds?: number
  file_size_bytes: number
  error_message?: string
  audio_metadata?: Record<string, unknown>
  speakers: Speaker[]
  created_at: string
  updated_at: string
}

export interface JobProgress {
  job_id: string
  stage: string
  progress: number
  message: string
}

export interface TranscriptSegment {
  id: string
  speaker_id: string
  speaker_label: string
  speaker_color: string
  start_time: number
  end_time: number
  text: string
  confidence: number
}

File: backend/app/models/job.py
import uuid
import enum
from typing import Optional, List, TYPE_CHECKING

from sqlalchemy import String, Integer, BigInteger, Float, Text, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID, JSONB

from app.database import Base
from app.models.base import TimestampMixin

if TYPE_CHECKING:
    from app.models.user import User
    from app.models.speaker import Speaker


class JobStatus(str, enum.Enum):
    UPLOADED = "UPLOADED"
    PREPROCESSING = "PREPROCESSING"
    VAD = "VAD"
    DIARIZING = "DIARIZING"
    EMBEDDING = "EMBEDDING"
    CLUSTERING = "CLUSTERING"
    SEPARATING = "SEPARATING"
    TRANSCRIBING = "TRANSCRIBING"
    ANALYZING = "ANALYZING"
    RECONSTRUCTING = "RECONSTRUCTING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class Job(TimestampMixin, Base):
    __tablename__ = "jobs"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    original_filename: Mapped[str] = mapped_column(String(512), nullable=False)
    original_s3_key: Mapped[str] = mapped_column(String(1024), nullable=False)
    processed_s3_key: Mapped[Optional[str]] = mapped_column(String(1024), nullable=True)
    status: Mapped[JobStatus] = mapped_column(
        SAEnum(JobStatus, name="job_status"),
        nullable=False,
        default=JobStatus.UPLOADED,
        index=True,
    )
    pipeline_stage: Mapped[Optional[str]] = mapped_column(String(64), nullable=True)
    progress: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    duration_seconds: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    file_size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False, default=0)
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    audio_metadata: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)

    user: Mapped["User"] = relationship("User", back_populates="jobs")
    speakers: Mapped[List["Speaker"]] = relationship(
        "Speaker", back_populates="job", cascade="all, delete-orphan", lazy="select"
    )
